core/arch/arm/plat-rcar/romapi.c

```diff
@@ -0,0 +1,139 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2021, EPAM Systems
+ */
+#include <kernel/panic.h>
+#include <mm/core_memprot.h>
+#include <mm/core_mmu.h>
+
+#include "rcar.h"
+#include "romapi.h"
+
+static int get_api_table_index(void)
+{
+	/*
+	 * Depending on SoC type and version, there are 4 possible addresses
+	 * for each ROMAPI function
+	 */
+	static int index __nex_data = -1;
+
+	if (index != -1)
+		return index;
+
+	switch (rcar_prr_value & PRR_PRODUCT_MASK) {
+	case PRR_PRODUCT_H3:
+		switch (rcar_prr_value & PRR_CUT_MASK) {
+		case PRR_CUT_10:	/* H3 ES1.0 */
+		case PRR_CUT_11:	/* H3 ES1.1 */
+			index = 0;
+			break;
+		case PRR_CUT_20:	/* H3 ES2.0 */
+			index = 1;
+			break;
+		default:	/* Newer H3 versions use unified table */
+			index = 3;
+			break;
+		}
+		break;
+	case PRR_PRODUCT_M3W:
+		switch (rcar_prr_value & PRR_CUT_MASK) {
+		case PRR_CUT_10:	/* M3 ES1.0 */
+			index = 2;
+			break;
+		default:	/* Newer M3 versions use unified table */
+			index = 3;
+			break;
+		}
+		break;
+	default:			/* All other SoCs use unified table */
+		index = 3;
+		break;
+	}
+
+	return index;
+}
+
+/* implemented in romapi_call.S */
+extern uint32_t __plat_call_romapi(paddr_t sp, paddr_t func, uint64_t arg1,
+				   uint64_t arg2, uint64_t arg3);
+
+static uint32_t plat_call_romapi(paddr_t func, uint64_t arg1,
+				 uint64_t arg2, uint64_t arg3)
+{
+	/*
+	 * If MMU is enabled, we need to use trampoline function that will
+	 * disable MMU and switch stack pointer to physical address. On other
+	 * hand, if MMU is disabled, we can call the ROM function directly.
+	 */
+	if (cpu_mmu_enabled()) {
+		uint32_t (*fptr)(paddr_t func, uint64_t arg1, uint64_t arg2,
+				 uint64_t arg3);
+
+		/*
+		 * __plat_call_romapi() is identity mapped and we should call it
+		 * at identity address, otherwise it will experience problem
+		 * when MMU will be disabled.
+		 */
+		fptr = (void *)virt_to_phys(__plat_call_romapi);
```

Comment:
Why is `virt_to_phys()` needed here?

Reply:
After relocation (if ASLR is enabled), `__plat_call_romapi` function will be mapped at two places: relocated and identity-mapped. But `__plat_call_romapi` symbol will point to relocated address. Correct me if I'm wrong.

I need to call this function by identity mapped address, because it will disable MMU for time being. `virt_to_phys` returns physical address, which is the same as virtual for identity-mapped memory.

Reply:
Ah, right, now that I read the comment again it makes sense. When I read this initially, for some reason I did not realize the function has two virtual addresses. I focused on the identity mapping and therefore the `virt_to_phys()` call looked like a no-op to me. I missed the fact that the symbol has the relocated value here. Perhaps add this to the comment for people like me...?

Reply:
Sure :)

I elaborated the comment. Also I addressed @jenswi-linaro's comments regarding cache line alignments and fixed a new issue with call to `dcache_op_all` from identity location.